Introduce native tracer support

frisitano opened the pull request:
This PR introduces native tracer support such that blocks can be prover using standard rpc methods. This is a replacement for https://github.com/0xPolygonZero/zero-bin/pull/51. This is currently blocked as alchemy is returning incorrect data for prestate tracer requests. I have opened a ticket with the support team and they are looking into it.

Merge steps:

1. Merge companion PR https://github.com/0xPolygonZero/zk_evm/pull/246
2. Repoint `zk_evm` dependencies in cargo to either the git repo or a new release.
3. Merge this PR into develop.

frisitano committed:
Introduce native tracer support

frisitano commented:
I have tested this over the block range of 19240650 - 19240850 and 199 blocks completed successfully with one failure. This yields a 99.5% success rate.

muursh commented:
> I have tested this over the block range of 19240650 - 19240850 and 199 blocks completed successfully with one failure. This yields a 99.5% success rate.

What was the failure? Which block and do you know why?

frisitano commented:
> > I have tested this over the block range of 19240650 - 19240850 and 199 blocks completed successfully with one failure. This yields a 99.5% success rate.
> 
> What was the failure? Which block and do you know why?

It's block 19240734. The error is:
```
Error: Failed to get proof for account: JsonRpcClientError(TimeoutError)
```

I have not had a chance to investigate the root cause of this but @Nashtare and I had concluded that the success rate was sufficiently high to merge as is and then leave this investigation / fix for a follow up issue / PR.

Nashtare commented on tools/prove_blocks.sh:
```diff
@@ -5,6 +5,7 @@
 # 2 --> End block index (inclusive)
 # 3 --> Rpc endpoint:port (eg. http://35.246.1.96:8545)
 # 4 --> Ignore previous proofs (boolean)
+# 5 --> Rpc type (eg. jerigon / native)
```
Can this be set before `Ignore previous proofs`? So that it goes in pair with the RPC endpoint, and we can keep the boolean as optional with default value. Not really blocking though

muursh commented on Cargo.toml:
```diff
@@ -21,9 +21,10 @@ keccak-hash = "0.10.0"
-evm_arithmetization = { git = "https://github.com/0xPolygonZero/zk_evm.git", tag = "v0.3.1" }
-trace_decoder = { git = "https://github.com/0xPolygonZero/zk_evm.git", tag = "v0.3.1" }
-proof_gen = { git = "https://github.com/0xPolygonZero/zk_evm.git", tag = "v0.3.1" }
+evm_arithmetization = { git = "https://github.com/fractal-zkp/zk_evm.git", branch = "trace-processing-new" }
```
We've merged your `zk_evm` PR now

frisitano replied:
Ok I'll integrate those changes.

frisitano replied:
I've updated to point to the develop branch of `zk_evm`. If we want to use a tag then we need to cut a new tag on `zk_evm`. It would be good if someone could test this against `jerigon` as I needed to make a small change to the jerigon code but I don't have access to the jerigon rpc.

muursh replied:
Yep we intend to do it next week.

Nashtare commented on rpc/src/rpc/native/mod.rs:
```diff
@@ -0,0 +1,113 @@
+            )
+        )?;
+
+        debug!("Got block result: {:?}", rpc_block_metadata.block_by_number);
+        debug!("Got trace result: {:?}", block_trace);
+        debug!("Got chain_id: {:?}", rpc_block_metadata.chain_id);
```
nit: maybe remove? Not sure this brings much

Nashtare commented on rpc/src/rpc/native/txn.rs:
```diff
@@ -0,0 +1,283 @@
+) -> Result<Vec<TxnInfo>> {
+    let mut futures_ordered = FuturesOrdered::new();
+
+    for tx_hash in &block.transactions {
```
Last time we tested blocks, I recall we needed to reverse the order here. Has this been tested since @frisitano?

frisitano replied:
Yes, I tested this on 200 blocks with a 99.5% success rate. The change was made below on line 30 in which we `push_back` instead of the prior `push_front`, see [ref](https://github.com/fractal-zkp/zero-bin/blob/5234bc9dc291750aa51b3090379acb9160fc57cf/rpc/src/rpc/native/txn.rs#L27-L33).

Nashtare replied:
Awesome, thanks!
The failure being due to the branch collapsing bug mentioned in the related issue I presume?

frisitano replied:
No that issue has been fixed (hacked) by fetching proofs after block execution and inserting short nodes (extension and leafs) into the trie builder. 

This failure is due to an error when trying to fetch a proof for an account - see https://github.com/0xPolygonZero/zero-bin/pull/81#issuecomment-2133679551. I haven't had an opportunity to look into it.

Nashtare commented on rpc/src/rpc/native/trie.rs:
```diff
@@ -0,0 +1,169 @@
+};
+
+/// The hash of an empty trie.
+pub const EMPTY_TRIE_HASH: H256 = H256([
+    0x56, 0xe8, 0x1f, 0x17, 0x1b, 0xcc, 0x55, 0xa6, 0xff, 0x83, 0x45, 0xe6, 0x92, 0xc0, 0xf8, 0x6e,
+    0x5b, 0x48, 0xe0, 0x1b, 0x99, 0x6c, 0xad, 0xc0, 0x01, 0x62, 0x2f, 0xb5, 0xe3, 0x63, 0xb4, 0x21,
+]);
```
non-blocking, as not exposed (yet) but we should try to group all these common constants somewhere (probably somewhere under `zk_evm` crates) and reuse/re-exports them instead of always redefining them

frisitano replied:
Agreed, I will add a TODO. I was also wondering if this module would be better positioned in the `zk_evm/mpt_trie` crate for reuse?

Nashtare replied:
yeah that'd make sense, it's a bit too specific on MPTs to be inside `zero-bin` I'd say.
cc @0xaatif @BGluth

BGluth replied:
Yeah I think inside `mpt_trie` probably makes the most sense.

frisitano commented:
I've opened a [PR](https://github.com/0xPolygonZero/zk_evm/pull/258) in `zk_evm` repo to migrate the `PartialTrieBuilder`. I've repointed deps of this PR to the new branch and also addressed the feedback. Once that PR is merged I will repoint deps and squash commits to make it ready for merge.

0xaatif commented on .gitignore:
```diff
@@ -13,3 +13,6 @@ proofs/
 verifier_state_*
+
+# System files
+.DS_Store
```
nit: newline

0xaatif commented on README.md:
```diff
@@ -85,6 +85,7 @@ Usage: leader [OPTIONS] <COMMAND>
 Commands:
   stdio    Reads input from stdin and writes output to stdout
   jerigon  Reads input from a Jerigon node and writes output to stdout
+  native   Reads input from a native node and writes output to stdout 
```
new-issue: CLI documentation in README should be autogenerated, else it'll just get out of date

0xaatif replied:
#92

0xaatif commented on rpc/src/cli.rs:
```diff
@@ -0,0 +1,64 @@
+                            .await?
+                    }
+                };
+                std::io::stdout().write_all(&serde_json::to_vec(&prover_input)?)?;
```
nonblocking: using this over `println` is super uncommon - why do this?

frisitano replied:
This is an artefact of what was there prior to starting this PR. I will change to `serde_json::to_writer_pretty(io::stdout(), &prover_input)?` which is currently being used in `develop`.

0xaatif replied:
> `println!` entails converting the input to a string

This is not true, there are no intermediate allocations.

Also the serde_json::to_writer call comes from yours truly, and again, avoids intermediate allocations

0xaatif replied:
> **Does the println macro allocate heap memory?** I.e. What's the difference (in terms of memory allocation) between calling the format macro to create a string to be printed by println and just formatting things in the println macro?
>
> No, print!, write! and other similar macros that do not return a String do not need to allocate memory; they call format_args! (which is a compiler built-in) to process the arguments and ultimately use std::fmt::write to **write data directly to its destination, without preparing a temporary String.**

Emphasis mine, from https://stackoverflow.com/a/57319140

You can prove this to yourself:
```rust
use core::fmt;

use clap::Parser;

#[derive(clap::Parser)]
struct Repeat {
    count: usize,
}

impl fmt::Display for Repeat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for _ in 0..self.count {
            f.write_str("repeat")?
        }
        Ok(())
    }
}

fn main() {
    println!("{}", Repeat::parse());
}
```

```console
$ /usr/bin/time -v ./target/debug/sketchbook 20000 2>&1 | rg 'Maximum resident set size'
#                                            ^^^^^ change this
```

RSS does not vary with input

0xaatif replied:
to explain, the above is semantically equivalent to:
```rust
let mut writer = std::io::stdout();
for _ in 0..repeat {
    <std::io::Stdout as std::io::Write>::write_all(&mut writer, b"repeat").unwrap()
}
```

But no-one writes that code, people just use `println!`, `writeln!` etc. 
You can imagine that avoiding intermediate allocations was an explicit goal of these APIs, given that they're used so often, often in environments without an allocator in the case of `core::fmt`

You may be confusing `println!` with `format!`, which _returns_ a `String`?

0xaatif replied:
> Buffering also has the advantage of batching and iteratively writing, rather than requiring a full allocation of an intermediate type (e.g., a string) before writing out in full.

This is also not true in this case, buffering is completely up to the OS for stdio. Notably, it varies if the output is a TTY or a pipe

0xaatif replied:
> I think we probably both agree that to_writer is the more performant solution, and the current std::io::stdout().write_all(&serde_json::to_vec(&prover_input)?)? is a bit odd.

Works for me, let's spend our time elsewhere

0xaatif commented on rpc/src/cli.rs:
```diff
@@ -0,0 +1,64 @@
+use url::Url;
+
+#[derive(Parser)]
+pub enum Cli {
```
nonblocking: why have a file with just a single struct? `main.rs` is hardly cluttered, and this module isn't shared outside the package

frisitano replied:
For the sake of modularity. I'm not strongly opinionated on this but I personally think it's cleaner. Happy to change if the consensus prefers to migrate this to main.

frisitano replied:
I've migrated cli back to main.

0xaatif commented on rpc/src/rpc/jerigon.rs:
```diff
@@ -0,0 +1,233 @@
+}
+
+#[async_trait]
+impl RpcClient for JerigonRpcClient {
```
I don't think a new trait is appropriate here, a different function for each branch would suffice.

frisitano replied:
My thinking was that in the future we will have some parent struct that will subscribe to new blocks and fetch the prover input using the rpc and invoke proving jobs. We would want this to be generic over the `RpcClient` hence then introduction of this trait. I can remove it.

0xaatif replied:
I think that such a parent struct will own an `impl Provider`, but we should defer that design thinking to the PR that implements it. I'd advocate for removal

frisitano replied:
removed the `async_trait`

0xaatif commented on rpc/src/rpc/jerigon.rs:
```diff
@@ -0,0 +1,233 @@
+
+#[derive(Debug)]
+pub struct JerigonRpcClient {
+    provider: Arc<RootProvider<Http<reqwest::Client>>>,
```
This shouldn't need an `Arc` - you've added one because `#[async_trait]` told you to, but the code would be cleaner if that trait didn't exist.

This is also misusing the `alloy_provider` types - code _should_ be generic over an `impl Provider` - you should never need to spell out the whole client except at instantiation.

frisitano replied:
updated

0xaatif commented on rpc/src/rpc/mod.rs:
```diff
@@ -0,0 +1,25 @@
+pub use native::NativeRpcClient;
+
+/// RPC CLIENT TRAIT
+/// ===============================================================================================
```
This makes rustdoc look weird

0xaatif commented on rpc/src/rpc/native/block.rs:
```diff
@@ -0,0 +1,34 @@
+
+use anyhow::{anyhow, Result};
+use ethers::providers::{Http, Provider, RetryClient};
+use ethers::types::H256;
```
we should NOT be using `ethers` since #78

frisitano replied:
removed use of ethers

0xaatif commented on rpc/src/rpc/native/block.rs:
```diff
@@ -0,0 +1,34 @@
+use anyhow::{anyhow, Result};
+use ethers::providers::{Http, Provider, RetryClient};
+use ethers::types::H256;
+use tokio::sync::Mutex;
```
async Mutex are almost always wrong or unnecessary

frisitano replied:
Could you elaborate on this please? Do you have any proposed change?

0xaatif replied:
> Contrary to popular belief, it is ok and often preferred to use the ordinary [Mutex](https://doc.rust-lang.org/nightly/std/sync/mutex/struct.Mutex.html) from the standard library in asynchronous code.

_[`tokio::sync::Mutex` docs](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use)_

They generally don't work well with cancellation, defy user expectations, maybe check out [this fragment](https://youtu.be/1zOd52_tUWg?si=kis16mw-8ReoEG5P&t=2003), or the above docs.

I think the most appropriate changes is to refactor the code so that no shared state is needed - can this be a map reduce instead?

frisitano replied:
> > Contrary to popular belief, it is ok and often preferred to use the ordinary [Mutex](https://doc.rust-lang.org/nightly/std/sync/mutex/struct.Mutex.html) from the standard library in asynchronous code.
> 
> _[`tokio::sync::Mutex` docs](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use)_
> 
> They generally don't work well with cancellation, defy user expectations, maybe check out [this fragment](https://youtu.be/1zOd52_tUWg?si=kis16mw-8ReoEG5P&t=2003), or the above docs.

Thanks for the insights.

> I think the most appropriate changes is to refactor the code so that no shared state is needed - can this be a map reduce instead?

Yes I can refactor this to be a map reduce.

frisitano replied:
updated to map reduce

0xaatif commented on rpc/src/rpc/native/metadata.rs:
```diff
@@ -0,0 +1,194 @@
+        hashes.push(
+            block
+                .hash
+                .ok_or_else(|| anyhow!("Block hash not present for block"))?,
```
```rust
use anyhow::Context as _;
```
```suggestion
                .context("Block hash not present for block")?,
```

0xaatif commented on rpc/src/rpc/native/metadata.rs:
```diff
@@ -0,0 +1,194 @@
+        let block_metadata = BlockMetadata {
+            block_beneficiary: block_by_number
+                .author
+                .ok_or_else(|| anyhow!("Block beneficiary not present for block"))?,
```
```suggestion
                .context("Block beneficiary not present for block")?,
```

You get the idea...

0xaatif commented on rpc/src/rpc/native/metadata.rs:
```diff
@@ -0,0 +1,194 @@
+}
+
+impl RpcBlockMetadata {
+    pub async fn fetch(
```
nonblocking: this would fit better as a top-level function.

0xaatif commented on rpc/src/rpc/native/metadata.rs:
```diff
@@ -0,0 +1,194 @@
+
+/// Fetches the chain id.
+async fn fetch_chain_id(provider: Arc<Provider<RetryClient<Http>>>) -> Result<U256> {
+    info!("Fetching chain id");
```
```suggestion
```

0xaatif commented on rpc/src/rpc/native/mod.rs:
```diff
@@ -0,0 +1,107 @@
+mod txn;
+
+/// NATIVE RPC CLIENT
+/// ===============================================================================================
```
as above

0xaatif commented on rpc/src/rpc/native/mod.rs:
```diff
@@ -0,0 +1,107 @@
+
+use anyhow::Result;
+use ethers::prelude::*;
+use ethers::types::GethDebugTracerType;
```
as above

0xaatif commented on rpc/src/cli.rs:
```diff
@@ -0,0 +1,64 @@
+use std::io::Write;
+
+use anyhow::Result;
```
nit: Because many crates have their own `error` and `result` types, it's common to keep them qualified.
i.e if I read type `Error`, does it mean `io::Error`, `serde_json::Error`... etc.
See e.g [here](https://docs.rs/tokio/latest/src/tokio/fs/file.rs.html#152)

`fn foo() -> anyhow::Result<T>` is more advisable in all cases.

This applies throughout this PR

0xaatif commented on rpc/src/rpc/native/mod.rs:
```diff
@@ -0,0 +1,107 @@
+
+/// A static retry policy that always retries.
+#[derive(Debug, Default)]
+pub struct StaticRetryPolicy;
```
Library code should never care about this - that's why we should be abstract over `Provider`

0xaatif commented on rpc/src/rpc/native/state.rs:
```diff
@@ -0,0 +1,195 @@
+        Vec<(H160, EIP1186ProofResponse)>,
+    ),
+    anyhow::Error,
+> {
```
nit
```suggestion
) -> anyhow::Result<[Vec<(H160, EIP1186ProofResponse)>; 2]> {
```

0xaatif commented on rpc/src/rpc/native/state.rs:
```diff
@@ -0,0 +1,195 @@
+                        keys.into_iter().collect(),
+                        Some((block_number - 1).into()),
+                    )
+                    .map_err(|e| anyhow!("Failed to get proof for account: {:?}", e))
```
```rust
use anyhow::Context as __;
```
```suggestion
                    .context("Failed to get proof for account")
```

Anyhow will preserve the source error in it's cause chain. This is the whole point of using anyhow

0xaatif commented on rpc/src/rpc/native/txn.rs:
```diff
@@ -0,0 +1,283 @@
+        futures::try_join!(tx_fut, tx_receipt_fut, pre_trace_fut, diff_trace_fut,)?;
+
+    Ok((
+        tx.ok_or_else(|| anyhow!("Transaction not found."))?,
```
```rust
use anyhow::Context as _;
```
```suggestion
        tx.context("Transaction not found.")?,
```

Also works on `Option`s

0xaatif commented on tools/debug_blocks.sh:
```diff
@@ -0,0 +1,27 @@
+    current_a=$((initial_a + i))
+    echo "Running debug block script with block=$current_a rpc=$b"
+    "$script_dir/debug_block.sh" $current_a $b $rpc_type
+done
```
nit: NL

0xaatif reviewed (changes requested):
The most important change is keeping with `alloy` from now on. 

I'd like to re-review after markups

frisitano commented:
> The most important change is keeping with `alloy` from now on.
> 
> I'd like to re-review after markups

I've been trying to migrate to alloy but I'm running into an issue when trying to rlp encode the `ReceiptEnvelope` to populate `trace_decoder::trace_protocol::TxnMeta`. I would expect the following to work but it doesn't:

```rust
    let tx_receipt = provider.get_transaction_receipt(tx_hash).await?.unwrap();
    let tx_receipt_rlp = tx_receipt.inner.encoded_2718();
```

If someone could take a look at this to assess what I may be doing wrong that would be appreciated.

0xaatif commented:
Does any of this help?
https://github.com/0xPolygonZero/eth-tx-proof/blob/3805968cf6c1d88fa40ca10a37e9bba00c05faf3/leader/src/lib.rs#L521-L527
```rust
/// There are two representations of domain objects in ethereum:
/// - RPC (JSON), in [`alloy::rpc`].
/// - RLP (binary), in [`alloy::consensus`].
///
/// This module provides best-effort [RLP encoding](alloy::rlp::Encodable) for
/// RPC types.
```

frisitano commented:
> Does any of this help?
> 
> https://github.com/0xPolygonZero/eth-tx-proof/blob/3805968cf6c1d88fa40ca10a37e9bba00c05faf3/leader/src/lib.rs#L521-L527
> 
> ```rust
> 
> /// There are two representations of domain objects in ethereum:
> 
> /// - RPC (JSON), in [`alloy::rpc`].
> 
> /// - RLP (binary), in [`alloy::consensus`].
> 
> ///
> 
> /// This module provides best-effort [RLP encoding](alloy::rlp::Encodable) for
> 
> /// RPC types.
> 
> ```

It certainly looks like it could, thanks!

frisitano commented:
I've implemented an initial migration to `alloy` but I have hit a blocker. Previously we were using a retry policy to mitigate the request failures due to rate limiting from rpc providers. `ethers` shipped with an out the box retry client. I have done some investigation to see how we can implement a retry policy for `alloy`. `alloy` does ship with a [`ClientBuilder`](https://github.com/alloy-rs/alloy/blob/f6ebef272c5bd2783bae3fe7ef6ebc1dc3cee015/crates/rpc-client/src/builder.rs#L18-L21) struct that wraps `tower::ServiceBuilder`. It is suggested that we should be able to introduce a retry layer using this builder however my efforts [(and those of others)](https://github.com/alloy-rs/alloy/issues/717) have not been successful thus far. Without the retry logic I am unable to test against rpc providers. 

This PR still needs a hygiene sweep but I would like to address the retry issue first.

frisitano commented:
This is ready for review along with https://github.com/0xPolygonZero/zk_evm/pull/258.

0xaatif commented on rpc/src/native/txn.rs:
```diff
@@ -0,0 +1,355 @@
+/// Processes the transactions in the given block and updates the code db.
+pub(super) async fn process_transactions<ProviderT, TransportT>(
+    block: &Block,
+    provider: Arc<ProviderT>,
```
not clear that you need an Arc to me

0xaatif commented on rpc/src/native/txn.rs:
```diff
@@ -0,0 +1,355 @@
+    ProviderT: Provider<TransportT>,
+    TransportT: Transport + Clone,
+{
+    let mut futures_ordered = FuturesOrdered::new();
```
you can collect straight into this instead of `push_back`-ing

0xaatif commented on rpc/src/native/txn.rs:
```diff
@@ -0,0 +1,355 @@
+    {
+        let provider = provider.clone();
+        futures_ordered
+            .push_back(async move { super::txn::process_transaction(provider, tx).await });
```
`async { foo().await }` can just be written as `foo()`. I'm sure clippy should pick up on this

frisitano replied:
I'm not sure the reason but clippy didn't report on this. Thanks for raising it, I've updated it.

0xaatif commented on rpc/Cargo.toml:
```diff
@@ -18,13 +18,18 @@ trace_decoder = { workspace = true }
 hex-literal = "0.4.1"
 itertools = "0.13.0"
 url = "2.5.0"
+primitive-types = "0.12.2"
```
I'd prefer a `__compat_primitive_types` name, as a speed bump until we've removed it entirely

0xaatif commented on rpc/src/compat.rs:
```diff
@@ -0,0 +1,69 @@
+pub trait ToPrimitive<Out> {
```
style: I think both of these traits could be a single `Compat` trait

0xaatif commented on rpc/src/native/mod.rs:
```diff
@@ -0,0 +1,32 @@
+    TransportT: Transport + Clone,
+{
+    let (block_trace, other_data) = try_join!(
+        block::process_block_trace(provider.clone(), block_number),
```
`&Provider: Provider`, so you don't need to clone here, and you don't need an `Arc`

0xaatif commented on rpc/src/metadata.rs:
```diff
@@ -0,0 +1,109 @@
+use alloy::{
```
style: one function per file?

0xaatif commented on rpc/src/native/block.rs:
```diff
@@ -0,0 +1,34 @@
+use std::sync::Arc;
```
style: one function per file?

0xaatif commented on rpc/src/native/mod.rs:
```diff
@@ -0,0 +1,32 @@
+use std::{collections::HashMap, sync::Arc};
```
style: one function per file?

0xaatif commented on rpc/src/retry.rs:
```diff
@@ -0,0 +1,148 @@
+use std::{
```
Worth adding a tracking issue for this as tech debt to be resolved when retries land upstream.

Sorry that you had to write/port all this.

frisitano replied:
No problem, it was a good learning experience. I've created an issue here: https://github.com/0xPolygonZero/zero-bin/issues/102.

0xaatif reviewed (approved):
could stand to be polished, no blockers

frisitano committed:
remove arc and refactor code

frisitano commented:
I've polished this PR as recommended. Before this is merged we need review and merge 
 https://github.com/0xPolygonZero/zk_evm/pull/258. I can then repoint the deps of this branch to the `zk_evm/develop` branch.

Nashtare commented:
@frisitano Should be ready for merging once you get latest `develop` in your branch (it already includes the version bump on `zk_evm`!)

frisitano commented:
Currently merging upstream develop.

atanmarko commented:
@frisitano Lindas https://github.com/0xPolygonZero/zero-bin/pull/101 also updated alloy dependencies, should we merge it first?

atanmarko commented on Cargo.toml:
```diff
@@ -26,9 +32,10 @@ alloy = { git = "https://github.com/alloy-rs/alloy", features = [
+evm_arithmetization = { git = "https://github.com/fractal-zkp/zk_evm.git", branch = "feat/partial_trie_builder" }
+mpt_trie = { git = "https://github.com/fractal-zkp/zk_evm.git", branch = "feat/partial_trie_builder" }
+trace_decoder = { git = "https://github.com/fractal-zkp/zk_evm.git", branch = "feat/partial_trie_builder" }
+proof_gen = { git = "https://github.com/fractal-zkp/zk_evm.git", branch = "feat/partial_trie_builder" }
```
You can use latest release here `v0.4.0`. As we will merge `zero-bin` and `zk_evm`, we will probably switch to current local repo path reference soon.

frisitano commented:
> @frisitano Lindas 0xPolygonZero/zero-bin#101 also updated alloy dependencies, should we merge it first?

Yes feel free to merge it first.

frisitano committed:
merge upstream

frisitano committed:
update README

frisitano committed:
remove unecessary tools

frisitano committed:
use reference for provider

atanmarko commented on rpc/src/retry.rs:
```diff
@@ -0,0 +1,148 @@
+use tower::{retry::Policy, Layer, Service};
+
+#[derive(Debug)]
+pub struct RetryPolicy {
```
Maybe we could use [backoff]( https://docs.rs/backoff/latest/backoff/) crate, no need to implement it ourselves?

atanmarko replied:
@frisitano Could I open a ticket to port our retry implementation to backoff crate? We may need backoff mechanism in various scenarios/modules, easy with external backoff library. Long term it would probably be easier than to maintain our implementation.

frisitano replied:
I had looked into a few options but they didn't fit into the `tower` stack which `alloy` uses under the hood. This current implementation appears to be the easiest / best solution for the moment. `alloy` are currently working on introducing a retry client upstream so we should migrate to that when it's ready. I've created an issue for this #102. See [here](https://github.com/alloy-rs/examples/pull/95/files) for retry implementation example.

frisitano replied:
> @frisitano Could I open a ticket to port our retry implementation to backoff crate? We may need backoff mechanism in various scenarios/modules, easy with external backoff library. Long term it would probably be easier than to maintain our implementation.

certainly, feel free 👍

frisitano commented:
I've completed the merge with `develop` and have tested `native`. It may be wise to test `jerigon` again but I don't have access to a `jergion` rpc.

atanmarko commented:
EDIT: obligatory parameter to select `jerigon` or `native` before `test_only` and `ignore previous proofs`

@frisitano You can probably update `tools/prove_jerigon.sh` to `prove_rpc.sh` with the default as jerigon client and optional parameter as `native`

frisitano commented:
> @frisitano You can probably update `tools/prove_jerigon.sh` to `prove_rpc.sh` with the default as jerigon client and optional parameter as `native`

Good catch! However, I don't think we will be able to make the rpc type optional as we can only have one optional argument (the last positional argument). Currently `test_only` is being used as the optional argument. I am happy to introduce it as a required argument, what do you think?

atanmarko commented:
Tested `prove_jerigon.sh` with `371ca61`, no regressions.

frisitano committed:
refactor prove_rpc tool

frisitano committed:
remove prove_jerigon tool

frisitano commented:
> EDIT: obligatory parameter to select `jerigon` or `native` before `test_only` and `ignore previous proofs`
> 
> @frisitano You can probably update `tools/prove_jerigon.sh` to `prove_rpc.sh` with the default as jerigon client and optional parameter as `native`

I've updated the script. Let me know your thoughts.

atanmarko commented:
> > EDIT: obligatory parameter to select `jerigon` or `native` before `test_only` and `ignore previous proofs`
> > @frisitano You can probably update `tools/prove_jerigon.sh` to `prove_rpc.sh` with the default as jerigon client and optional parameter as `native`
> 
> I've updated the script. Let me know your thoughts.

@frisitano Script is OK. Please fix also the docs and the CI.

frisitano committed:
update README for prove_rpc.sh script changes

frisitano commented:
> > > EDIT: obligatory parameter to select `jerigon` or `native` before `test_only` and `ignore previous proofs`
> > > @frisitano You can probably update `tools/prove_jerigon.sh` to `prove_rpc.sh` with the default as jerigon client and optional parameter as `native`
> > 
> > 
> > I've updated the script. Let me know your thoughts.
> 
> @frisitano Script is OK. Please fix also the docs and the CI.

README updated. As discussed on slack, no changes required to CI.

atanmarko commented on rpc/src/jerigon.rs:
```diff
@@ -0,0 +1,100 @@
+}
+
+/// Obtain the prover input for a given block interval
+pub async fn prover_input<ProviderT, TransportT>(
```
@frisitano This one is duplicate, could be removed?

frisitano replied:
good catch. Yes looks like I'd made the change locally but didn't save the file. Updating now.

frisitano replied:
done

frisitano committed:
remove dead code

atanmarko commented:
@frisitano We want to fix alloy version, so we will use recently released tag [v0.1.1](https://github.com/alloy-rs/alloy/commit/f04a704ac5be9cb21b4b1f6c35f925607f1aa939).
Here is the [PR](https://github.com/fractal-zkp/zero-bin/pull/1). Could you please regression test it with native tracer? Jerigon works without regressions with v0.1.1.

atanmarko commented:
@Nashtare @BGluth Any additional comments/requests or we are ready to merge?

Nashtare reviewed (approved):
@atanmarko Only did a synthetic review over the latest changes but I'm good to merge.

frisitano commented:
> @frisitano We want to fix alloy version, so we will use recently released tag [v0.1.1](https://github.com/alloy-rs/alloy/commit/f04a704ac5be9cb21b4b1f6c35f925607f1aa939). Here is the [PR](https://github.com/fractal-zkp/zero-bin/pull/1). Could you please regression test it with native tracer? Jerigon works without regressions with v0.1.1.

Regression tested latest develop with the native tracer. No regressions, working as expected.
